### ISSUE TITLE
[RFC] Materialise Range

### ISSUE THREAD

**Issue by larsk-db** [RoaringBitmap/RoaringBitmap#479]
I would like some feedback on how the community feels about adding the following kind of API:

> Materialise a sub-range of the bitmap's "contains" space into a pre-allocated array.

That is, I am proposing an API equivalent to the following code on a `RoaringBitmap bitmap`, for arbitrary values of `int BUFFER_SIZE`, `int startIndex`:
```java
boolean[] batch = new boolean[BUFFER_SIZE];
for (int pos = 0; pos < BUFFER_SIZE; pos++) {
	batch[pos] = bitmap.contains(startIndex + pos);
}
```

### Why is this relevant?

There are applications, where both the `true` and `false` values on the `contains`-method within the universe of a bitmap matter. But iterator implementations focus only on the `true` values, and force us to infer the `false` values from gaps in the sequence of contained values.

#### Example
Consider a paginated table view where one column is logically equivalent to the "contains" value of the bitmap, i.e. the column's value for each `row` is derived from `bitmap.contains(row)`. The bitmap contains the values for the whole table, but we only want to show this table by materialising one page at a time. We might even skip pages, if the user jumps forward, so we wouldn't want to materialize the bitmap for pages that are never loaded.

### Why do we need a special API for this?

The code presented above works just fine, but the main concern is performance. If we consider the table example above, but simply push it a few levels lower in the execution stack, where we are trying to execute vectorised operations on the column(s), it becomes clear why performance is an important concern here.

While random lookups on `RoaringBitmap` aren't crazy expensive, we can do much better if we can push filling subranges of the buffer array into the individual container types. In particular for `RunContainer` (but also to some degree for `BitmapContainer`) we can get much better performance using APIs like `Arrays.fill` instead of setting a single position at a time.
The performance argument gets even stronger for `Roaring64Bitmap`, where the tree traversal is much more expensive for random lookups.

We can do a bit better than random lookups performance-wise using (batched) iterators with `advanceIfNeeded`, but it is still not ideal. In particular we still materialise the `int` (`RoaringBitmap`) or `long` (`Roaring64Bitmap`) values contained in the bitmap from their parts, when really we only want to know which "contains"-equivalent value should go into which position of the `buffer`.

### What do you mean by "contains"-equivalent?

Basically, I would prefer the API not to be limited to materialising only `true` and `false` values produced by `contains`, but actually arbitrary values `containsValue`, `notContainsValue` of an arbitrary type `$type` (primitive or object) to the following code (or using a mapper function instead):
```java
$type[] batch = new $type[BUFFER_SIZE];
for (int pos = 0; pos < BUFFER_SIZE; pos++) {
	if (bitmap.contains(startIndex + pos)) {
		batch[pos] = containsValue;
	} else {
		batch[pos] = notContainsValue;
	}
}
```
Baking this capability directly into the API would avoid an additional array allocation to transform the buffer from `boolean` to the desired target type. It would also again benefit from range setings via `Arrays.fill`.

### Option 1: Function call per range

The top level API could look like:

```java
public <V> void materializeRangeInto(
            long startIndex,
            V[] buffer,
            V containsValue,
            V notContainsValue)
```

The disadvantage of this variant is that the generic type parameter doesn't work well with primitives, which are the most likely values to be used here, imo. If people are against materialising custom values in favour of only `boolean`, then this API is probably the simplest and fastest.

### Option2: Visitor API

The top level API could consist of a function like:

```java
public void materializeRangeInto(
	long startIndex,
	MaterializeVisitor visitor);
```

and a visitor interface like:

```java
public interface MaterializeVisitor {
	void set(int relativePos);
	void unset(int relativePos);
	void setAll(int relativeFrom, int relativeTo);
	void unsetAll(int relativeFrom, int relativeTo);
}
```
the `relative*` arguments here are always to the `startIndex`, and thus correspond exactly to positions in the underlying buffer. The contract would be that there may be no "gaps" in the sequence of indices invoking the visitor, so that pre-initialisation of the underlying buffer is not required. In particular, this would allow the same buffer to be re-used for multiple invocations, assuming its contents have been processed/consumed already.

This API would allow users to write their own specialized versions for the primitives they care about and we could provide a generic example implementation that can be used for all object types (and boxed primitves, if someone really wants to).

Another nice thing about the visitor API is, that it acutally allows different indexable datastructures than just raw arrays to be filled. It could be off-heap data, or a memory-mapped file, for example. This makes it a very powerful and generic API, that I believe could serve other uses-cases than the one I presented above as well.

## Implemenation

If there is consensus on a design, I'd be happy to provide a PR with an implementation. I already have a hacky prototype for Option 1, which I could complete or convert to Option 2 or something agreeable API.
I just don't want to put more effort into this, before getting some feedback, so my work will not be in vain :)

**Comment by lemire** [comment]
@larsk-db 

Here is my proposal.

Can you come up with a "proof of concept" benchmark to illustrate the performance benefits you foresee? Take a problem, implement it with the current API. Measure the performance. Then guess how much faster it could get, and report that. Whatever you are trying to do... I would start by trying to look at what the current API allows. If it is not 'slow' then it is probably not worth the effort. So we want something that the current API does not allow efficiently.

> Basically, I would prefer the API not to be limited to materialising only true and false values produced by contains, but actually arbitrary values containsValue, notContainsValue of an arbitrary type $type (primitive or object) to the following code (or using a mapper function instead):

Something like

```Java
myclass[] array = new myclass[BUFFER_SIZE];
for (int pos = 0; pos < BUFFER_SIZE; pos++) {
	batch[pos] = bitmap.contains(startIndex + pos) ? myclass.containsValue : notContainsValue;
}
```

feels like a performance anti-pattern. Just the array makes me cringe. You are going to be spending 64 bits per possible value in your range just for the array, even assuming that the `myclass` instance are free.

I'd work very hard to avoid writing such code.


> The code presented above works just fine, but the main concern is performance. 

The following code is indeed not great (performance-wise):

```Java
boolean[] batch = new boolean[BUFFER_SIZE];
for (int pos = 0; pos < BUFFER_SIZE; pos++) {
	batch[pos] = bitmap.contains(startIndex + pos);
}
```

But you can do much better with iterators... something like this...

```Java
PeekableIntIterator j = // get an iterator
j.advanceIfNeeded(startIndex);
while ( j.hasNext() ) {
  int x =  j.next();
  if( x is beyond your scope ) {
   break;
  }
  batch[x] = true;
}
```

The larger point is that we have iterators that allow you to iterate over all values inside a range.


We have other fancy iterators like batch iterators.

We have other fancy methods like `selectRangeWithoutCopy` that could be used as well.

> we can do much better if we can push filling subranges of the buffer array into the individual container types. In particular for RunContainer (but also to some degree for BitmapContainer) we can get much better performance using APIs like Arrays.fill instead of setting a single position at a time.


Though that is true, you have to make sure that you are benchmarking any new proposal against the best use of the current RoaringBitmap API, with its fast iterators and other methods.

It would be sad to work super hard on a new API only to find out that the pre-existing iterators get you 99% there.

Also, if you can prove that something meaningful is 'slow', then it motivates the work.

**Comment by larsk-db** [comment]
@lemire 

> Can you come up with a "proof of concept" benchmark to illustrate the performance benefits you foresee? Take a problem, implement it with the current API. Measure the performance. Then guess how much faster it could get, and report that. Whatever you are trying to do... I would start by trying to look at what the current API allows. If it is not 'slow' then it is probably not worth the effort. So we want something that the current API does not allow efficiently.

So, I have done this within the context of my use-case, and there I could observe between a 3x and 150x performance difference between the iterator implementation and the materialise implementation, and between 100x and 700x between random lookup and materialise. The range here depends on the size of the filter. The iterator implementation does better on smaller filters, because its seek isn't great; all of this is for `Roaring64Bitmap`, which is what I really care about. I assume for the 32-bit variants the differences are going to be much smaller.

Now these numbers do include some of the interaction with our propriety code, so I can't really share the benchmark. Do you feel it's necessary as motivation (or also for later reference) that I port a variant of this benchmark to open code I can share?
 
> Something like
> 
> ```java
> myclass[] array = new myclass[BUFFER_SIZE];
> for (int pos = 0; pos < BUFFER_SIZE; pos++) {
> 	batch[pos] = bitmap.contains(startIndex + pos) ? myclass.containsValue : notContainsValue;
> }
> ```
> 
> feels like a performance anti-pattern. Just the array makes me cringe. You are going to be spending 64 bits per possible value in your range just for the array, even assuming that the `myclass` instance are free.
> 
> I'd work very hard to avoid writing such code.

Why do you think so? If you take my example from above with the table, the `myclass` could be `String` and `containsValue` and `notContainsValue` could be two interned strings like `"Inlcuded"` `"Not Included"`, for example. That would be a pretty reasonable usage of a class in this context, with essentially no allocation necessary. And if I'm materialising "pages" of say 4k entries at a time, I'm not spending more than 32KB in storage, which seems pretty reasonable to me.

I do agree that primitives are more likely as a target type here, which is why I specifically didn't like the generic API of Option 1 that doesn't support unboxed primitives without specialising for every single one. The benchmark results above, for example, are with using a `byte` array there, which is actually a pretty fast representation of a `boolean` array, if one is willing to spend the extra memory space.


> 
> > The code presented above works just fine, but the main concern is performance.
> 
> The following code is indeed not great (performance-wise):
> 
> ```java
> boolean[] batch = new boolean[BUFFER_SIZE];
> for (int pos = 0; pos < BUFFER_SIZE; pos++) {
> 	batch[pos] = bitmap.contains(startIndex + pos);
> }
> ```
> 
> But you can do much better with iterators... something like this...
> 
> ```java
> PeekableIntIterator j = // get an iterator
> j.advanceIfNeeded(startIndex);
> while ( j.hasNext() ) {
>   int x =  j.next();
>   if( x is beyond your scope ) {
>    break;
>   }
>   batch[x] = true;
> }
> ```
> 
> The larger point is that we have iterators that allow you to iterate over all values inside a range.
> 
> We have other fancy iterators like batch iterators.
> 
> We have other fancy methods like `selectRangeWithoutCopy` that could be used as well.

Yes, as I said, iterators are definitely a possible implementation for this. 
A few points here, though they may slightly digress:
- Most of the fancy iterators are only available for the 32-bit bitmap. Roaring64Bitmap has only gotten support even for seek after I actually did the benchmarks (I actually wrote my own for the benchmarks, so the results of the benchmarks above are essentially for the `advanceIfNeeded` that was added in https://github.com/RoaringBitmap/RoaringBitmap/commit/8ec2c9b36e65a4b95177ff6e713b2cdd3d6a14d9 since I kinda implemented the same thing, just without reverse support.)
- Even with the seek implementation we now have on `Roaring64Bitmap` materialisation still gets slower with larger bitmaps, because instead of jumping to the right leaf and scanning from there, we actually search along the leaves. This certainly a lot cheaper than searching along the actual values, but it still isn't great for bitmaps with a few million entries or more. In fact my benchmarks showed that random lookups become faster than this seek when the bitmap hits around 64million entries (and we do materialisation in the middle). Fixing this is kind of orthogonal to this issue and on my TODO list.
- My strongest performance point against using the iterators is still that they assemble the `int` (32-bit) or `long` (64-bit) values inside the bitmap, which I don't really care about. I only need to know which values goes into which relative position in the buffer. This probably isn't a biggie for the 32-bit version, but for the 64-bit I've seen flamegraphs indicating that we spend almost 1/3rd of the compute time just assembling those `long` values, which matches up pretty well with the 3x performance improvement I've mentioned above.

> Though that is true, you have to make sure that you are benchmarking any new proposal against the best use of the current RoaringBitmap API, with its fast iterators and other methods.
> 
> It would be sad to work super hard on a new API only to find out that the pre-existing iterators get you 99% there.
> 
> Also, if you can prove that something meaningful is 'slow', then it motivates the work.

Absolutely agreed. So let me know if the benchmarks I mentioned are sufficient motivation in your eyes, or you'd prefer for me to port them, so you can try to optimise the implementation yourself etc.

**Comment by richardstartin** [comment]
For pagination you can figure out the index of the page (i.e. `(i >>> PAGE_SHIFT)`) and the index within the page (i.e. `(i & PAGE_MASK)`) and track when the page index has changed whilst doing inverted iteration.

```java

class PaginatingConsumer<T> implements LongConsumer {

static final int PAGE_SIZE = 4096;
static final int PAGE_MASK = PAGE_SIZE - 1;
static final int PAGE_SHIFT = Integer.bitCount(PAGE_MASK);

    private final T[] page = makePage();
    private long currentPageIndex = -1;
    private boolean dirty = false;

    @Override
    public void accept(long i) {
        long pageIndex = i >>> PAGE_SHIFT;
        if (pageIndex != currentPageIndex && dirty) {
            flush();
            resetPage();
            currentPageIndex = pageIndex;
            dirty = false;
        }
        page[index] = indexToValue((int)(i & PAGE_MASK));
        dirty = true;
    }

    T indexToValue(int i) {
       // stuff
    }

    void resetPage() {
       // e.g. Arrays.fill(page, defaultValue);
    }

    void flush() {
       if (dirty) {
         // do stuff with the page here
       }
    }

}
...
PaginatingConsumer<T> consumer = makeNewPaginatingConsumer();
bitmap.forEach(consumer);
consumer.flush();
```

You would skip over potentially millions of empty pages in very sparse bitmaps, but that feels like it's a feature not a bug.

I also appreciate that inverted iteration can be viral and doesn't compose well. 

For more efficient code so that whatever it is you're doing with the indexes can potentially be vectorised and for bounds checks to get eliminated, adding a callback to `LongConsumer` for runs would be handy, at the cost of no longer being a `@FunctionalInterface`:

```java
interface LongConsumer {

void accept(long index);

// write a for loop in here but be careful not to use a long induction variable
void acceptRun(long min, long max);

}
```

**Comment by larsk-db** [comment]
@richardstartin Thanks, I've tried something like this approach, too, actually. The problem is that `bitmap.forEach` always goes through the whole bitmap, so it's the bitmap that controls when which page gets materialised. I need an external entity (say the user, or a data processing executor) to control which page should get materialised, based on what it wants to see/process right now. So more of a "pull-based" approach, rather than "push", you could say.

**Comment by lemire** [comment]
@larsk-db 


This looks reasonable:

```Java
public interface MaterializeVisitor {
	void set(int relativePos);
	void unset(int relativePos);
	void setAll(int relativeFrom, int relativeTo);
	void unsetAll(int relativeFrom, int relativeTo);
}
```

We can probably debate a bit the names. Are you sure about `MaterializeVisitor`? What you have there is a consumer. Whether it 'materializes' anything is up to you.

What about `RangeConsumer` or something?

Note that your consumer is explicitly 32-bit which might be totally fine. The `materializeRangeInto` probably needs a final index as well, or (maybe better) a length parameter of some kind.

We can discuss the name too.

I realize that you want to materialize ranges, but that's not what these functions do. That's what they allow you to do.

> The problem is that bitmap.forEach always goes through the whole bitmap, so it's the bitmap that controls when which page gets materialised

Of course, we could implement `forEach(..., long start, long end)` or something like it which has the benefit of being generic and serving a lot of people well. It is more or less what you end up doing in any case.

The big difference is that you are 'run' aware.

> My strongest performance point against using the iterators is still that they assemble the int (32-bit) or long (64-bit) values inside the bitmap, which I don't really care about. 

On the one hand, casting a 32-bit integer to a 64-bit integer is quite cheap and may even be optimized away if you are subtracting from a base. On the other hand, if you have 64-bit bitmaps, then you paid the heavy price once by going to a 64-bit bitmap, instead of using 32-bit fragments which is the performance choice.

> Why do you think so? 

One can represent an array of Boolean values in a bitset using 1 bit per value, or in array using, say, 64 bits per value. There is a 64x memory effect + the loss of word-level parallelism.

This becomes irrelevant if we go for your `MaterializeVisitor` because I do not have to know where or how you store the data.

> Could you point me to this? I can't seem to find anything in the docs or using Github's (admittedly terrible) search.

The `selectRangeWithoutCopy` function is not directly exposed but it is used to narrow down some operations (say `and`) over a range.

**Comment by larsk-db** [comment]
> @larsk-db
> 
> This looks reasonable:
> 
> ```java
> public interface MaterializeVisitor {
> 	void set(int relativePos);
> 	void unset(int relativePos);
> 	void setAll(int relativeFrom, int relativeTo);
> 	void unsetAll(int relativeFrom, int relativeTo);
> }
> ```
> 
> We can probably debate a bit the names. Are you sure about `MaterializeVisitor`? What you have there is a consumer. Whether it 'materializes' anything is up to you.
> 
> What about `RangeConsumer` or something?

Ah, yes fair point. I was thinking about visitor pattern originally, because it was doing tree traversal and I wanted to do different things per container. But since it's not actually exposing the different container types, I do concur that `RangeConsumer` is a more correct name.

> 
> Note that your consumer is explicitly 32-bit which might be totally fine. The `materializeRangeInto` probably needs a final index as well, or (maybe better) a length parameter of some kind.

For the version with an array as parameter it was pretty clear from the length of the array what the end index should be. But for the Visitor/Consumer API, we definitely need to provide a length or end index. I don't really have a preference for either, since they are so easy to convert.

As for the 32-bit part, my thinking when I wrote that was still that we'd be only materialising into arrays and there relative indices can't be any larger anyway. The idea with memory mapped files, for example, which would be long addressable only popped into my mind later.

I see two options here:
1. We make an `IntRangeConsumer` and a `LongRangeConsumer`, just like we have now with `IntConsumer` and `LongConsumer`.
2. We only build a 32-bit consumer and enforce this limit by making the length argument discussed above be an `int`. If someone really needs to materialise more than a full int space at a time, they can invoke the function multiple times with sequential offsets.

Personally, I'm kinda into 2., because materialising more than 2billion values, even if they are just 1 byte each, seems like an edge case. But then, as you said, the consumer API is not necessarily only useful for materialisation, so 1. might be a very valid choice.

> 
> We can discuss the name too.
> 
> I realize that you want to materialize ranges, but that's not what these functions do. That's what they allow you to do.
> 
> > The problem is that bitmap.forEach always goes through the whole bitmap, so it's the bitmap that controls when which page gets materialised
> 
> Of course, we could implement `forEach(..., long start, long end)` or something like it which has the benefit of being generic and serving a lot of people well. It is more or less what you end up doing in any case.
> 
> The big difference is that you are 'run' aware.

We could combine the `RangeConsumer` with the `forEach(..., long start, long end)` and make a specialised version for `LongConsumer` that just creates and adapter to `RangeConsumer` (by converting `setAll` to an `accept` for each value, and ignoring `unset` and `unsetAll`) and then invokes the more general implementation. This way we give a lot of flexibility in the use of the API in a way, that I *think* the JIT compiler should be pretty good at optimising away.

I do think we should give the version that takes a `RangeConsumer` a more different sounding name, though, to account for the offset based logic and the fact that something will be invoked both for present and absent values.

Which actually gives me an idea for better naming in the interface. How about the following?

```java
public interface RelativeRangeConsumer {
  void acceptPresent(int relativePos);
  void acceptAbsent(int relativePos);
  void acceptAllPresent(int relativeFrom, int relativeTo);
  void acceptAllAbsent(int relativeFrom, int relativeTo);
}
```

> On the one hand, casting a 32-bit integer to a 64-bit integer is quite cheap and may even be optimized away if you are subtracting from a base. On the other hand, if you have 64-bit bitmaps, then you paid the heavy price once by going to a 64-bit bitmap, instead of using 32-bit fragments which is the performance choice.

Oh yeah, the cast would be very cheap. What isn't cheap is `LongUtils.toLong`. Compare that to the 32-bit relative to startingIndex math where we just keep the highBits in an int and for every offset just use a single operation to add the low bits. I don't have a choice in my use of `Roaring64Bitmap`, since I don't necessarily control the size of the data entered into the bitmap in the first place and it can be rather large. But we can avoid this cost for this particular case, if we choose an API that only allows 32-bit relative offset based addressing.

 
> One can represent an array of Boolean values in a bitset using 1 bit per value, or in array using, say, 64 bits per value. There is a 64x memory effect + the loss of word-level parallelism.

Yes one can. But an efficient storage format isn't helpful if you can't do the work you need to do at the end of the pipeline, because you can decompress it efficiently to the format you actually need to work with.

> This becomes irrelevant if we go for your `MaterializeVisitor` because I do not have to know where or how you store the data.

True, and it gives people a lot of flexibility to work with whatever materialised format they want and more. In fact in my own use-case the target format isn't really a simple array either. We simply already have code that is efficient at using an array as a basis for a column, so I thought it would make for a good API. But after this discussion, I'm actually convinced that the Visitor/Consumer thing would be much better, both in general and for us, because we could actually materialise directly into the target data-structure.

> The `selectRangeWithoutCopy` function is not directly exposed but it is used to narrow down some operations (say `and`) over a range.

Ah, I found it now. Should have searched in IntelliJ directly. Github search is truly terrible. 
Looking at it, it still just a lot of work to remove values from containers at the boundaries. Which makes sense when your selections are typically larger than 1 container and the boundaries are the exception. But at least for my use-case it'd be the other way around. Most invocations would stay within a single container and crossing a boundary would be the exception.
Or at least that is my preferred implementation. I could, of course, over-read significantly to amortise the tree traversals and then later do range copies of an already larger materialised array for every page.

**Comment by lemire** [comment]
Quick comment: I agree. No need for a 64-bit version. I was not criticizing your proposal. It is perfectly fine to have a 32-bit consumer on a 64-bit data structure. Of course, it needs to be documented (which can be done with a single sentence). I also agree with `RelativeRangeConsumer`. It sounds good.

If someone needs a 64-bit consumer, we can add it later.

**Comment by larsk-db** [comment]
Forgot to close this. It's addressed in #492
